From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Ambrus Toth <palpista11@gmail.com>
Date: Sat, 3 Oct 2026 23:31:05 +0200
Subject: [PATCH 14/14] Build and run tests like on Linux on Darwin

nixpkgs uses swift-corelibs-xctest on Darwin because XCTest.framework
from Xcode is not available. Like on Linux, it requires test products to
be executables with a generated entry point instead of bundles that are
loaded by the `xctest` tool from Xcode.

Build test products as executables on Darwin as well, and run them
directly like on Linux. This allows XCTest-based tests to run on Darwin.
Swift Testing-based tests run the same executable.
---
 Sources/Build/BuildPlan/BuildPlan+Test.swift  |  4 ++--
 Sources/Commands/SwiftTestCommand.swift       | 22 -------------------
 Sources/PackageModel/SwiftSDKs/SwiftSDK.swift |  7 +++---
 .../BuildParameters+Testing.swift             |  3 ++-
 4 files changed, 7 insertions(+), 29 deletions(-)

diff --git a/Sources/Build/BuildPlan/BuildPlan+Test.swift b/Sources/Build/BuildPlan/BuildPlan+Test.swift
index 545ecb2..ba27cde 100644
--- a/Sources/Build/BuildPlan/BuildPlan+Test.swift
+++ b/Sources/Build/BuildPlan/BuildPlan+Test.swift
@@ -188,7 +188,7 @@ extension BuildPlan {
             let swiftTargetDependencies: [Module.Dependency]
             let resolvedTargetDependencies: [ResolvedModule.Dependency]
 
-            if destinationBuildParameters.triple.isDarwin() {
+            if case .loadableBundle = destinationBuildParameters.testProductStyle {
                 discoveryTargets = nil
                 swiftTargetDependencies = []
                 resolvedTargetDependencies = []
@@ -198,7 +198,7 @@ extension BuildPlan {
                 resolvedTargetDependencies = [.module(discoveryTargets!.resolved, conditions: [])]
             }
 
-            if !destinationBuildParameters.triple.isDarwin(), let entryPointResolvedTarget = testProduct.testEntryPointModule {
+            if case .entryPointExecutable = destinationBuildParameters.testProductStyle, let entryPointResolvedTarget = testProduct.testEntryPointModule {
                 if isEntryPointPathSpecifiedExplicitly || explicitlyEnabledDiscovery {
                     if isEntryPointPathSpecifiedExplicitly {
                         // Allow using the explicitly-specified test entry point target, but still perform test discovery and thus declare a dependency on the discovery modules.
diff --git a/Sources/Commands/SwiftTestCommand.swift b/Sources/Commands/SwiftTestCommand.swift
index c2ae2ad..2eefae5 100644
--- a/Sources/Commands/SwiftTestCommand.swift
+++ b/Sources/Commands/SwiftTestCommand.swift
@@ -282,15 +282,12 @@ public struct SwiftTestCommand: AsyncSwiftCommand {
             // Validate XCTest is available on Darwin-based systems. If it's not available and we're hitting this code
             // path, that means the developer must have explicitly passed --enable-xctest (or the toolchain is
             // corrupt, I suppose.)
-            let toolchain = try swiftCommandState.getTargetToolchain()
             if case let .unsupported(reason) = try swiftCommandState.getHostToolchain().swiftSDK.xctestSupport {
                 if let reason {
                     throw TestError.xctestNotAvailable(reason: reason)
                 } else {
                     throw TestError.xcodeNotInstalled
                 }
-            } else if toolchain.targetTriple.isDarwin() && toolchain.xctestPath == nil {
-                throw TestError.xcodeNotInstalled
             }
 
             if !self.options.shouldRunInParallel {
@@ -900,11 +897,7 @@ final class TestRunner {
             testSpecifier = testSpecifiers.joined(separator: ",")
         }
 
-#if os(macOS)
-        return ["-XCTest", testSpecifier]
-#else
         return [testSpecifier]
-#endif
     }
 
     /// Creates an instance of TestRunner.
@@ -965,23 +958,8 @@ final class TestRunner {
             args.append(testPath.relative(to: localFileSystem.currentWorkingDirectory!).pathString)
             args.append(contentsOf: self.additionalArguments)
         } else {
-#if os(macOS)
-            switch library {
-            case .xctest:
-                guard let xctestPath = self.toolchain.xctestPath else {
-                    throw TestError.xcodeNotInstalled
-                }
-                args += [xctestPath.pathString]
-            case .swiftTesting:
-                let helper = try self.toolchain.getSwiftTestingHelper()
-                args += [helper.pathString, "--test-bundle-path", testPath.pathString]
-            }
-            args += self.additionalArguments
-            args += [testPath.pathString]
-    #else
             args += [testPath.pathString]
             args += self.additionalArguments
-    #endif
         }
 
         if library == .swiftTesting {
diff --git a/Sources/PackageModel/SwiftSDKs/SwiftSDK.swift b/Sources/PackageModel/SwiftSDKs/SwiftSDK.swift
index e0e75b9..6ee52c4 100644
--- a/Sources/PackageModel/SwiftSDKs/SwiftSDK.swift
+++ b/Sources/PackageModel/SwiftSDKs/SwiftSDK.swift
@@ -597,10 +597,9 @@ public struct SwiftSDK: Equatable {
             extraSwiftCFlags.append(contentsOf: sdkPaths.frameworks.flatMap { ["-F", $0.pathString] })
             extraSwiftCFlags.append(contentsOf: sdkPaths.libraries.flatMap { ["-I", $0.pathString] })
             extraSwiftCFlags.append(contentsOf: sdkPaths.libraries.flatMap { ["-L", $0.pathString] })
-            xctestSupport = .supported
-        } catch {
-            xctestSupport = .unsupported(reason: String(describing: error))
-        }
+        } catch {}
+        // nixpkgs uses swift-corelibs-xctest on Darwin, which is part of the toolchain.
+        xctestSupport = .supported
         #else
         xctestSupport = .supported
         #endif
diff --git a/Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift b/Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift
index 357c7de..2872581 100644
--- a/Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift
+++ b/Sources/SPMBuildCore/BuildParameters/BuildParameters+Testing.swift
@@ -119,7 +119,8 @@ extension BuildParameters {
 
     /// The style of test product to produce.
     public var testProductStyle: TestProductStyle {
-        return triple.isDarwin() ? .loadableBundle : .entryPointExecutable(
+        // nixpkgs uses swift-corelibs-xctest on Darwin, which works like on other platforms.
+        return .entryPointExecutable(
             explicitlyEnabledDiscovery: testingParameters.explicitlyEnabledDiscovery,
             explicitlySpecifiedPath: testingParameters.explicitlySpecifiedPath
         )
